README.md
[![Build Status](https://travis-ci.org/ambitioninc/fleming.png)](https://travis-ci.org/ambitioninc/fleming)
Fleming
================

This repository contains the Fleming package, which contains a set of routines for doing datetime manipulation. Named after Sandford Fleming, the father of worldwide standard timezones, this package is meant to aid datetime manipulations with regards to timezones.

Fleming addresses some of the common difficulties with timezones and datetime objects, such as performing arithmetic and datetime truncation across a Daylight Savings Time border. It also provides utilities for generating date ranges and getting unix times with respect to timezones.

Fleming accepts pytz timezone objects as parameters, and it is assumed that the user has a basic understanding of pytz. Click [here](http://pytz.sourceforge.net/) for more information about pytz.

## Installation
To install the latest release, type:

    pip install fleming

To install the latest code directly from source, type:

    pip install git+git://github.com/ambitioninc/fleming.git

## Function Overview
A brief description of each function in this package is below. More detailed descriptions and advanced usage of the functions follow after that. Click on the function names to go to their detailed descriptions.
- [convert_to_tz](#convert_to_tz): Converts a datetime object into a provided timezone.
- [add_timedelta](#add_timedelta): Adds a timedelta to a datetime object.
- [floor](#floor): Rounds a datetime object down to the previous time interval.
- [ceil](#ceil): Rounds a datetime object up to the next time interval.
- [intervals](#intervals): Gets a range of times at a given timedelta interval.
- [unix_time](#unix_time): Returns a unix time stamp of a datetime object.

Note that all of these functions properly handle Daylight Savings Time transitions and other artifacts not normally supported in datetime manipulation. Keep reading for more detailed descriptions and examples of the functions.

### convert_to_tz(dt, tz, return_naive=False)<a name="convert_to_tz"></a>
Given an aware or naive datetime dt, convert it to the timezone tz.

**Args:**
- dt: A naive or aware datetime object. If dt is naive, it has UTC set as its timezone.
- tz: A pytz timezone object specifying the timezone to which dt should be converted.
- return_naive: A boolean describing whether the return value should be a naive datetime object.

**Returns:**
An aware datetime object that was the result of converting dt into tz. If return_naive is True, the returned value has no tzinfo set.

**Examples:**

    import fleming
    import datetime
    import pytz

    dt = datetime.datetime(2013, 2, 4)
    print dt
    2013-02-04 00:00:00

    # Convert naive UTC time to aware EST time
    dt = fleming.convert_to_tz(dt, pytz.timezone('US/Eastern'))
    print dt
    2013-02-03 19:00:00-05:00

    # Convert aware EST time to aware CST time
    dt = fleming.convert_to_tz(dt, pytz.timezone('US/Central'))
    print dt
    2013-02-03 18:00:00-06:00

    # Convert aware CST time back to naive UTC time
    dt = fleming.convert_to_tz(dt, pytz.utc, return_naive=True)
    print dt
    2013-02-04 00:00:00

### add_timedelta(dt, td, within_tz=None)<a name="add_timedelta"></a>
Given a naive or aware datetime dt, add a timedelta td to it and return it. If within_tz is specified, the datetime arithmetic happens with regard to the timezone. Proper measures are used to ensure that datetime arithmetic across a DST border is handled properly.

**Args:**
- dt: A naive or aware datetime object. If it is naive, it is assumed to be UTC.
- td: A timedelta (or relativedelta) object to add to dt.
- within_tz: A pytz timezone object. If provided, dt will be converted to this timezone before datetime arithmetic and then converted back to its original timezone afterwards.

**Returns:**
A datetime object that results from adding td to dt. The timezone of the returned datetime will be equivalent to the original timezone of dt (or its DST equivalent if a DST border was crossed). If the original time was naive, the returned value is naive.

**Examples:**

    import pytz
    import datetime
    import fleming

    # Do a basic timedelta addition to a naive UTC time and create a naive UTC time
    # two weeks in the future
    dt = datetime.datetime(2013, 3, 1)
    dt = fleming.add_timedelta(dt, datetime.timedelta(weeks=2))
    print dt
    2013-03-15 00:00:00

    # Do addition on an EST datetime where the arithmetic does not cross over DST
    dt = fleming.convert_to_tz(dt, pytz.timezone('US/Eastern'))
    print dt
    2013-03-14 20:00:00-04:00

    dt = fleming.add_timedelta(dt, datetime.timedelta(weeks=2, days=1))
    print dt
    2013-03-29 20:00:00-04:00

    # Do timedelta arithmetic such that it starts in DST and crosses over into no DST.
    # Note that the hours stay in tact and the timezone changes
    dt = fleming.add_timedelta(dt, datetime.timedelta(weeks=-4))
    print dt
    2013-03-01 20:00:00-05:00

    # Take an aware UTC time and do datetime arithmetic in regards to EST. Do the arithmetic
    # such that a DST border is crossed
    dt = datetime.datetime(2013, 3, 1, 5, tzinfo=pytz.utc)
    # It should be midnight in EST
    print fleming.convert_to_tz(dt, pytz.timezone('US/Eastern'))
    2013-03-01 00:00:00-05:00

    # Do arithmetic on the UTC time with respect to EST.
    dt = fleming.add_timedelta(
        dt, datetime.timedelta(weeks=2), within_tz=pytz.timezone('US/Eastern'))
    # The hour (4) of the returned UTC time is different that the original (5).
    print dt
    2013-03-15 04:00:00+00:00

    # However, the hours in EST still reflect midnight
    print fleming.convert_to_tz(dt, pytz.timezone('US/Eastern'))
    2013-03-15 00:00:00-04:00


### floor(dt, within_tz=None, year=None, month=None, week=None, day=None, hour=None, minute=None, second=None, microsecond=None)<a name="floor"></a>
Perform a floor on a datetime, rounding the datetime to its nearest provided interval. Available intervals are year, month, week, day, hour, minute, second, and microsecond.

For example, to round to the nearest month, provide month=1 as input. Give a value like 3 to the month argument and it will round to the nearest trimonth in a year (i.e. the nearest quarter). Values for other intervals operate in the same way, such as rounding down to the nearest day in a month (or nearest triday).

The only paramter to this function which is not like the others is the week parameter. The week parameter rounds the time down to its nearest Monday. In contrast to other intervals, week can only be 1 since it has no larger time interval from which to start a tri or quadweek for example.

Also, if the week parameter is provided, the year and month arguments are also not used in the floor calculation.

Note that multiple combinations of attributes can be used where they make sense, such as flooring to the nearest trimonth and triday (month=3, day=3).

**Args:**
- dt: A naive or aware datetime object. If it is naive, it is assumed to be UTC
- within_tz: A pytz timezone object. If given, the floor will be performed with respect to the timezone.
- year: Specifies the yearly interval to round down to. Defaults to None.
- month: Specifies the monthly interval (inside of a year) to round down to. Defaults to None.
- week: Specifies to round to the beginning of the previous week. Defaults to None and only accepts a possible value of 1.
- day: Specifies the daily interval to round down to (inside of a month). Defaults to None.
- hour: Specifies the hourly interval to round down to (inside of a day). Defaults to None.
- minute: Specifies the minute interval to round down to (inside of an hour). Defaults to None.
- second: Specifies the second interval to round down to (inside of a minute). Defaults to None.
- microsecond: Specfies the microsecond interval to round down to (inside of a second). Defaults to None.

**Returns:**
A datetime object that results from flooring dt to the interval. The timezone of the returned datetime will be equivalent to the original timezone of dt (or its DST equivalent if a DST border was crossed). If the input time was naive, it returns a naive datetime object.

**Raises:**
ValueError if the interval is an invalid value.

**Examples:**

    import datetime
    import pytz
    import fleming

    # Do basic floors in naive UTC time. Results are naive UTC.
    print fleming.floor(datetime.datetime(2013, 3, 3, 5), year=1)
    2013-01-01 00:00:00
    print fleming.floor(datetime.datetime(2013, 3, 3, 5), month=1)
    2013-03-01 00:00:00

    # Weeks start on Monday, so the floor will be for the previous Monday
    print fleming.floor(datetime.datetime(2013, 3, 3, 5), week=1)
    2013-02-25 00:00:00

    print fleming.floor(datetime.datetime(2013, 3, 3, 5), day=1)
    2013-03-03 00:00:00

    # Pass an aware datetime and return an aware datetime
    print fleming.floor(
        datetime.datetime(2013, 3, 3, 5, tzinfo=pytz.utc), day=1)
    2013-03-03 00:00:00+00:00

    # Peform a floor in EST. The result is in EST
    dt = fleming.convert_to_tz(
        datetime.datetime(2013, 3, 4, 6), pytz.timezone('US/Eastern'))
    print dt
    2013-03-04 01:00:00-05:00

    print fleming.floor(dt, year=1)
    2013-01-01 00:00:00-05:00

    print fleming.floor(dt, day=1)
    2013-03-04 00:00:00-05:00

    # Now perform a floor that starts out of DST and ends up in DST. The
    # timezones before and after the floor will be different, but the
    # time values are correct
    dt = fleming.convert_to_tz(
        datetime.datetime(2013, 11, 28, 6), pytz.timezone('US/Eastern'))
    print dt
    2013-11-28 01:00:00-05:00

    print fleming.floor(dt, month=1)
    2013-11-01 00:00:00-04:00

    # Start with a naive UTC time and floor it with respect to EST
    dt = datetime.datetime(2013, 2, 1)
    # Since it is January 31 in EST, the resulting floored value
    # for a day will be the previous day. Also, the returned value is
    # in the original naive timezone of UTC
    print fleming.floor(dt, day=1, within_tz=pytz.timezone('US/Eastern'))
    2013-01-31 00:00:00

    # Similarly, EST values can be floored relative to CST values.
    dt = fleming.convert_to_tz(
        datetime.datetime(2013, 2, 1, 5), pytz.timezone('US/Eastern'))
    print dt
    2013-02-01 00:00:00-05:00

    # Since it is January 31 in CST, the resulting floored value
    # for a day will be the previous day. Also, the returned value is
    # in the original timezone of EST
    print fleming.floor(dt, day=1, within_tz=pytz.timezone('US/Central'))
    2013-01-31 00:00:00-05:00

    # Get the starting of a quarter by using month=3
    print fleming.floor(datetime.datetime(2013, 2, 4), month=3)
    2013-01-01 00:00:00


### ceil(dt, within_tz=None, year=None, month=None, week=None, day=None, hour=None, minute=None, second=None, microsecond=None)<a name="ceil"></a>
Perform a ceil on a datetime to the next closest interval in the future. For example, if month=1, this function will round up the time to the next month in the future. Larger numbers can be used, such as month=3, to round up to the beginning of the next quarter.

Note that this function allows combinations of interval variables (such as month=2 and day=2 to round up to the next duomonth of the year and next duoday of the month), but the smaller intervals are always not important since they will always be at the beginning of the larger interval.

**Args:**
- dt: A naive or aware datetime object. If it is naive, it is assumed to be UTC.
- within_tz: A pytz timezone object. If given, the ceil will be performed with respect to the timezone.
- year: Specifies the yearly interval to round up to. Defaults to None.
- month: Specifies the monthly interval (inside of a year) to round up to. Defaults to None.
- week: Specifies to round up to the beginning of the next week. Defaults to None and only accepts a possible value of 1.
- day: Specifies the daily interval to round up to (inside of a month). Defaults to None.
- hour: Specifies the hourly interval to round up to (inside of a day). Defaults to None.
- minute: Specifies the minute interval to round up to (inside of an hour). Defaults to None.
- second: Specifies the second interval to round up to (inside of a minute). Defaults to None.
- microsecond: Specfies the microsecond interval to round up to (inside of a second). Defaults to None.

**Returns:**
A datetime object that results from ceiling dt to the next interval. The timezone of the returned datetime will be equivalent to the original timezone of dt (or its DST equivalent if a DST border was crossed). If the original datetime object was naive, the returned object is naive.

**Raises:**
ValueError if the interval is not a valid value.

**Examples:**

    import datetime
    import pytz
    import fleming

    # Do basic ceils in naive UTC time. Results are naive UTC
    print fleming.ceil(datetime.datetime(2013, 3, 3, 5), year=1)
    2014-01-01 00:00:00

    print fleming.ceil(datetime.datetime(2013, 3, 3, 5), month=1)
    2013-04-01 00:00:00

    # Weeks start on Monday, so the floor will be for the next Monday
    print fleming.ceil(datetime.datetime(2013, 3, 3, 5), week=1)
    2013-03-04 00:00:00

    print fleming.ceil(datetime.datetime(2013, 3, 3, 5), day=1)
    2013-03-04 00:00:00

    # Use aware datetimes to return aware datetimes
    print fleming.ceil(
        datetime.datetime(2013, 3, 3, 5, tzinfo=pytz.utc), day=1)
    2013-03-04 00:00:00+00:00

    # Peform a ceil in CST. The result is in Pacfic time
    dt = fleming.convert_to_tz(
        datetime.datetime(2013, 3, 4, 7), pytz.timezone('US/Pacific'))
    print dt
    2013-03-03 23:00:00-08:00

    print fleming.ceil(dt, year=1)
    2014-01-01 00:00:00-08:00

    print fleming.ceil(dt, day=1)
    2013-03-04 00:00:00-08:00

    # Do a ceil with respect to EST. Since it is March 4 in EST, the
    # returned value is March 5 in Pacific time
    print fleming.ceil(dt, day=1, within_tz=pytz.timezone('US/Eastern'))
    2013-03-05 00:00:00-08:00

    # Note that doing a ceiling on a time that is already on the boundary
    # returns the original time
    print fleming.ceil(datetime.datetime(2013, 4, 1), month=1)
    2013-04-01 00:00:00


### intervals(start_dt, td, within_tz=None, stop_dt=None, is_stop_dt_inclusive=False, count=0)<a name="intervals"></a>
Returns a range of datetime objects starting from start_dt and going in increments of timedelta td. If stop_dt is specified, the intervals go to stop_dt (and include stop_dt in the return if is_stop_dt_inclusive=True). If stop_dt is None, the count variable is used to control how many iterations are in the time intervals. If stop_dt is None and count is None, a generator will be returned that can yield any number of datetime objects.

**Args:**
- start_dt: A naive or aware datetime object from which to start the time intervals. If it is naive, it is assumed to be UTC.
- td: A timedelta object describing the time interval in the intervals.
- within_tz: A pytz timezone object. If provided, the intervals will be computed with respect to this timezone.
- stop_dt: A naive or aware datetime object that specifies the end of the intervals. Defaults to being exclusive in the intervals. If naive, it is assumed to be in UTC.
- is_stop_dt_inclusive: True if the stop_dt should be included in the time intervals. Defaults to False.
- count: If set, an integer specifying a count of intervals to use if stop_dt is None.  If stop_dt is None and count is None, a generator will be returned that can yield any number of datetime objects. Defaults to None.

**Returns:**
A generator of datetime objects. The datetime objects are in the original timezone of the start_dt (or its DST equivalent if a border is crossed). If the input is naive, the returned intervals are naive.

**Examples:**

    import datetime
    import pytz
    import fleming

    # Using a naive UTC time, get intervals of time for every day.
    for dt in fleming.intervals(datetime.datetime(2013, 2, 3), datetime.timedelta(days=1), count=5):
        print dt
    2013-02-03 00:00:00
    2013-02-04 00:00:00
    2013-02-05 00:00:00
    2013-02-06 00:00:00
    2013-02-07 00:00:00

    # Use an EST time. Do intervals of a day. Cross the DST time border on March 10th.
    est_dt = fleming.convert_to_tz(datetime.datetime(2013, 3, 9, 5), pytz.timezone('US/Eastern'))
    for dt in fleming.intervals(est_dt, datetime.timedelta(days=1), count=5):
        print dt
    2013-03-09 00:00:00-05:00
    2013-03-10 00:00:00-05:00
    2013-03-11 00:00:00-04:00
    2013-03-12 00:00:00-04:00
    2013-03-13 00:00:00-04:00

    # Similarly, we can iterate through UTC times while doing the date range with respect to EST. Note
    # that the UTC hour changes as the DST border is crossed on March 10th.
    for dt in fleming.intervals(
            datetime.datetime(2013, 3, 9, 5), datetime.timedelta(days=1), within_tz=pytz.timezone('US/Eastern'),
            count=5):
        print dt
    2013-03-09 05:00:00
    2013-03-10 05:00:00
    2013-03-11 04:00:00
    2013-03-12 04:00:00
    2013-03-13 04:00:00

    # If we don't specify a count or stop time, we can iterate indefinitely.
    for dt in fleming.intervals(datetime.datetime(2013, 1, 1), datetime.timedelta(days=1)):
        print dt
    2013-01-01 00:00:00
    2013-01-02 00:00:00
    2013-01-03 00:00:00
    ....
    2013-05-05 00:00:00
    ....
    to the end of time...

    # Use a stop time. Note that the stop time is exclusive
    for dt in fleming.intervals(
            datetime.datetime(2013, 3, 9), datetime.timedelta(weeks=1), stop_dt=datetime.datetime(2013, 3, 23)):
        print dt
    2013-03-09 00:00:00
    2013-03-16 00:00:00


docs/release_notes.rst
Release Notes
=============

v0.7.1
------
* Read the Docs config file v2

v0.7.0
------
* Add support for python 3.9, 3.10
* Drop support for python 3.5, 3.6

v0.6.0
------
* Add support for python 3.8
* Drop python 2.7

v0.5.0
------
* Remove support for python 3.3 and 3.4
* Add support for python 3.7

v0.4.6
------
* Officially support python 3.6

v0.4.5
------
* Fixed a bug that cropped up when trying to floor times to quarter boundaries
* Fixed a bug that caused tests to fail on machines having non-UTC timezones

v0.4.4
------
* Officially support python 3.5

v0.4.1
------
* Added sphinx documentation and converted all docs to RestructuredText


flemming/flemming.py
"""
Provides helpers for manipulating datetime objects with respect to time zones.

Authors:
    Wes Kendall (https://github.com/wesleykendall)
    Jeff McRiffey (https://github.com/jmcriffey)
    Josh Marlow (https://github.com/joshmarlow)
"""
import datetime

import pytz


def attach_tz_if_none(dt, tz):
    """Makes a naive timezone aware or returns it if it is already aware.

    Attaches the timezone tz to the datetime dt if dt has no tzinfo. If
    dt already has tzinfo set, return dt.

    Args:
        dt: A naive or aware datetime object.
        tz: A pytz timezone object.

    Returns:
        An aware datetime dt with tzinfo set to tz. If dt already had tzinfo
        set, dt is returned unchanged.
    """
    return tz.localize(dt) if dt.tzinfo is None else dt


def remove_tz_if_return_naive(dt, return_naive):
    """A helper function to strip tzinfo objects if return_naive is True.

    Returns a naive datetime object if return_naive is True.

    Args:
        dt: A datetime object.
        return_naive: A boolean.

    Returns:
        A naive dt if return_naive is True.
    """
    return dt.replace(tzinfo=None) if return_naive else dt


def convert_to_tz(dt, tz, return_naive=False):
    """Converts a time into another timezone.

    Given an aware or naive datetime dt, convert it to the timezone tz.

    Args:
        dt: A naive or aware datetime object. If dt is naive, it has UTC
            set as its timezone.
        tz: A pytz timezone object specifying the timezone to which dt
            should be converted.
        return_naive: A boolean describing whether the return value
            should be a naive datetime object.

    Returns:
        An aware datetime object that was the result of converting dt
        into tz. If return_naive is True, the returned value has no
        tzinfo set.

    Examples:
        import flemming
        import datetime
        import pytz

        dt = datetime.datetime(2013, 2, 4)
        print dt
        2013-02-04 00:00:00

        # Convert naive UTC time to aware EST time
        dt = flemming.convert_to_tz(dt, pytz.timezone('US/Eastern'))
        print dt
        2013-02-03 19:00:00-05:00

        # Convert aware EST time to aware CST time
        dt = flemming.convert_to_tz(dt, pytz.timezone('US/Central'))
        print dt
        2013-02-03 18:00:00-06:00

        # Convert aware CST time back to naive UTC time
        dt = flemming.convert_to_tz(dt, pytz.utc, return_naive=True)
        print dt
        2013-02-04 00:00:00
    """
    return remove_tz_if_return_naive(
        tz.normalize(attach_tz_if_none(dt, pytz.utc)), return_naive)


def dst_normalize(dt):
    """Normalizes a datetime that crossed a DST border to its new DST timezone.

    Given an aware datetime object dt, call pytz's normalize function on dt's timezone
    If dt had crossed a DST border because of datetime arithmetic, its timezone will
    be changed to reflect the new DST timezone (for example, going from EST to EDT).
    The original time values from dt remain unchanged.

    An example input to this function would be.
    t = 2013/03/11:05:00:00 EST
    Since march 11 overlaps DST, this datetime object is really in EDT. The returned value would be
    t = 2013/03/11:05:00:00 EDT.

    Args:
        dt: An aware datetime object.

    Returns:
        An aware datetime object that has time value identical to dt. The only possible difference
        is that its timezone might have been converted into a DST timezone.
    """
    return dt.replace(tzinfo=convert_to_tz(dt, dt.tzinfo).tzinfo)


def add_timedelta(dt, td, within_tz=None, return_naive=False):
    """Adds a timedelta to a datetime. Can add timedeltas relative to a timezone.

    Given a naive or aware datetime dt, add a timedelta td to it and return it. If
    within_tz is specified, the datetime arithmetic happens with regard to the timezone.
    Proper measures are used to ensure that datetime arithmetic across a DST border
    is handled properly.

    Args:
        dt: A naive or aware datetime object. If it is naive, it is assumed to be UTC.
        td: A timedelta (or relativedelta) object to add to dt.
        within_tz: A pytz timezone object. If provided, dt will be converted to this
            timezone before datetime arithmetic and then converted back to its original
            timezone afterwards.
        return_naive: A boolean defaulting to False. If True, the result is returned
            as a naive datetime object with tzinfo equal to None.

    Returns:
        An aware datetime object that results from adding td to dt. The timezone of the
        returned datetime will be equivalent to the original timezone of dt (or its DST
        equivalent if a DST border was crossed). If return_naive is True, the returned
        value has no tzinfo object.

    Examples:
        import pytz
        import datetime
        import flemming

        # Do a basic timedelta addition to a naive UTC time and create an aware UTC time
        # two weeks in the future
        dt = datetime.datetime(2013, 3, 1)
        dt = flemming.add_timedelta(dt, datetime.timedelta(weeks=2))
        print dt
        2013-03-15 00:00:00+00:00

        # Do addition on an EST datetime where the arithmetic does not cross over DST
        dt = flemming.convert_to_tz(dt, pytz.timezone('US/Eastern'))
        print dt
        2013-03-14 20:00:00-04:00

        dt = flemming.add_timedelta(dt, datetime.timedelta(weeks=2, days=1))
        print dt
        2013-03-29 20:00:00-04:00

        # Do timedelta arithmetic such that it starts in DST and crosses over into no DST.
        # Note that the hours stay in tact and the timezone changes
        dt = flemming.add_timedelta(dt, datetime.timedelta(weeks=-4))
        print dt
        2013-03-01 20:00:00-05:00

        # Take a UTC time and do datetime arithmetic in regards to EST. Do the arithmetic
        # such that a DST border is crossed
        dt = datetime.datetime(2013, 3, 1, 5)
        # It should be midnight in EST
        print flemming.convert_to_tz(dt, pytz.timezone('US/Eastern'))
        2013-03-01 00:00:00-05:00

        # Do arithmetic on the UTC time with respect to EST.
        dt = flemming.add_timedelta(
            dt, datetime.timedelta(weeks=2), within_tz=pytz.timezone('US/Eastern'))
        # The hour (4) of the returned UTC time is different that the original (5).
        print dt
        2013-03-15 04:00:00+00:00

        # However, the hours in EST still reflect midnight
        print flemming.convert_to_tz(dt, pytz.timezone('US/Eastern'))
        2013-03-15 00:00:00-04:00
    """
    # Make sure it is aware
    dt = attach_tz_if_none(dt, pytz.utc)
    # Localize it to time_zone if within_tz is specified
    loc_dt = convert_to_tz(dt, within_tz) if within_tz else dt
    # Add the time delta in the localized time zone, taking care of any DST border crossings
    loc_dt = dst_normalize(loc_dt + td)
    # Convert it back to the original timezone if within_tz was specified
    loc_dt = convert_to_tz(loc_dt, dt.tzinfo) if within_tz else loc_dt
    # Return it, stripping the timezone information if return_naive
    return remove_tz_if_return_naive(loc_dt, return_naive)


def floor(dt, floor, within_tz=None, return_naive=False):
    """Floors a datetime to the nearest time boundary.

    Perform a 'floor' on a datetime, where the floor variable can be:
    'year', 'month', 'day', 'hour', 'minute', 'second', or 'week'. This
    function will round the datetime down to the beginning of the start
    of the floor.

    Args:
        dt: A naive or aware datetime object. If it is naive, it is
            assumed to be UTC
        floor: The interval to be floored. Can be 'year', 'month',
            'week', 'day', 'hour', 'minute', or 'second'.
        within_tz: A pytz timezone object. If given, the floor will
            be performed with respect to the timezone.
        return_naive: A boolean specifying whether to return the
            datetime object as naive.

    Returns:
        An aware datetime object that results from flooring dt to floor. The timezone of the
        returned datetime will be equivalent to the original timezone of dt (or its DST
        equivalent if a DST border was crossed). If return_naive is True, the returned
        value has no tzinfo object.

    Raises:
        ValueError if floor is not a valid floor value.

    Examples:
        import datetime
        import pytz
        import flemming

        # Do basic floors in naive UTC time. Results are UTC aware
        print flemming.floor(datetime.datetime(2013, 3, 3, 5), 'year')
        2013-01-01 00:00:00+00:00

        print flemming.floor(datetime.datetime(2013, 3, 3, 5), 'month')
        2013-03-01 00:00:00+00:00

        # Weeks start on Monday, so the floor will be for the previous Monday
        print flemming.floor(datetime.datetime(2013, 3, 3, 5), 'week')
        2013-02-25 00:00:00+00:00

        print flemming.floor(datetime.datetime(2013, 3, 3, 5), 'day')
        2013-03-03 00:00:00+00:00

        # Use return_naive if you don't want to return aware datetimes
        print flemming.floor(
            datetime.datetime(2013, 3, 3, 5), 'day', return_naive=True)
        2013-03-03 00:00:00

        # Peform a floor in EST. The result is in EST
        dt = flemming.convert_to_tz(
            datetime.datetime(2013, 3, 4, 6), pytz.timezone('US/Eastern'))
        print dt
        2013-03-04 01:00:00-05:00

        print flemming.floor(dt, 'year')
        2013-01-01 00:00:00-05:00

        print flemming.floor(dt, 'day')
        2013-03-04 00:00:00-05:00

        # Now perform a floor that starts out of DST and ends up in DST. The
        # timezones before and after the floor will be different, but the
        # time values are correct
        dt = flemming.convert_to_tz(
            datetime.datetime(2013, 11, 28, 6), pytz.timezone('US/Eastern'))
        print dt
        2013-11-28 01:00:00-05:00

        print flemming.floor(dt, 'month')
        2013-11-01 00:00:00-04:00

        # Start with a naive UTC time and floor it with respect to EST
        dt = datetime.datetime(2013, 2, 1)
        # Since it is January 31 in EST, the resulting floored value
        # for a day will be the previous day. Also, the returned value is
        # in the original timezone of UTC
        print flemming.floor(dt, 'day', within_tz=pytz.timezone('US/Eastern'))
        2013-01-31 00:00:00+00:00

        # Similarly, EST values can be floored relative to CST values.
        dt = flemming.convert_to_tz(
            datetime.datetime(2013, 2, 1, 5), pytz.timezone('US/Eastern'))
        print dt
        2013-02-01 00:00:00-05:00

        # Since it is January 31 in CST, the resulting floored value
        # for a day will be the previous day. Also, the returned value is
        # in the original timezone of EST
        print flemming.floor(dt, 'day', within_tz=pytz.timezone('US/Central'))
        2013-01-31 00:00:00-05:00
    """
    # Make sure it is aware
    dt = attach_tz_if_none(dt, pytz.utc)
    # Localize it to time_zone if within_tz is specified
    loc_dt = convert_to_tz(dt, within_tz) if within_tz else dt

    # Floor the local time to its proper interval
    if floor == 'year':
        loc_dt = loc_dt.replace(month=1, day=1, hour=0, minute=0, second=0, microsecond=0)
    elif floor == 'month':
        loc_dt = loc_dt.replace(day=1, hour=0, minute=0, second=0, microsecond=0)
    elif floor == 'day':
        loc_dt = loc_dt.replace(hour=0, minute=0, second=0, microsecond=0)
    elif floor == 'hour':
        loc_dt = loc_dt.replace(minute=0, second=0, microsecond=0)
    elif floor == 'minute':
        loc_dt = loc_dt.replace(second=0, microsecond=0)
    elif floor == 'second':
        loc_dt = loc_dt.replace(microsecond=0)
    elif floor == 'week':
        # Floor a week to a day and then do proper arithmetic to get to the beginning of week
        loc_dt = loc_dt.replace(hour=0, minute=0, second=0, microsecond=0)
        loc_dt -= datetime.timedelta(days=loc_dt.weekday())
    else:
        raise ValueError('{0} is not a valid floor value'.format(floor))

    # Set the timezone back to that of the original time. Do a DST normalization in case any
    # DST boundaries were crossed
    loc_dt = dst_normalize(loc_dt.replace(tzinfo=dt.tzinfo))

    # Return it, stripping the timezone information if return_naive
    return remove_tz_if_return_naive(loc_dt, return_naive)


def intervals(
        start_dt, td, within_tz=None, stop_dt=None, is_stop_dt_inclusive=False, count=0,
        return_naive=False):
    """Returns a range of datetime objects with a timedelta interval.

    Returns a range of datetime objects starting from start_dt and going in increments of
    timedelta td. If stop_dt is specified, the intervals go to stop_dt (and include
    stop_dt in the return if is_stop_dt_inclusive=True). If stop_dt is None, the
    count variable is used to control how many iterations are in the time intervals.

    Args:
        start_dt: A naive or aware datetime object from which to start the time intervals.
            If it is naive, it is assumed to be UTC.
        td: A timedelta object describing the time interval in the intervals.
        within_tz: A pytz timezone object. If provided, the intervals will be computed with
            respect to this timezone.
        stop_dt: A naive or aware datetime object that specifies the end of the intervals.
            Defaults to being exclusive in the intervals. If naive, it is assumed to
            be in UTC.
        is_stop_dt_inclusive: True if the stop_dt should be included in the time
            intervals. Defaults to False.
        count: An integer specifying a count of intervals to use if stop_dt is None.
        return_naive: All datetimes in the intervals are returned as naive objects.

    Returns:
        A list of datetime objects.
    """
    # Make sure start_dt is aware
    start_dt = attach_tz_if_none(start_dt, pytz.utc)

    # Make sure the stop_dt is aware
    stop_dt = attach_tz_if_none(stop_dt, pytz.utc) if stop_dt is not None else None

    # Set initial time and loop iteration values
    time_iter = start_dt
    loop_counter = 0

    while True:
        # Break when the end criterion has been met
        if ((stop_dt is None and loop_counter >= count) or
                (stop_dt is not None and is_stop_dt_inclusive and time_iter > stop_dt) or
                (stop_dt is not None and not is_stop_dt_inclusive and time_iter >= stop_dt)):
            break

        yield remove_tz_if_return_naive(time_iter, return_naive)

        # Increment the time iteration and the loop counter
        time_iter = add_timedelta(time_iter, td, within_tz=within_tz)
        loop_counter += 1


def unix_time(dt, within_tz=None, return_ms=False):
    """Converts a datetime object to a unix timestamp.

    Converts a naive or aware datetime object to unix timestamp. If within_tz
    is present, the timestamp returned is relative to that time zone.

    Args:
        dt: A naive or aware datetime object. If it is naive,
            it is assumed to be UTC.
        within_tz: A pytz timezone object if the user wishes to
            return the unix time relative to another timezone.
        return_ms: A boolean specifying to return the value
            in milliseconds since the Unix epoch. Defaults to False.

    Returns:
        An integer timestamp since the Unix epoch. If return_ms is
        True, returns the timestamp in milliseconds.

    Examples:
        import datetime
        import pytz
        import flemming

        # Do a basic naive UTC conversion
        dt = datetime.datetime(2013, 4, 2)
        print flemming.unix_time(dt)
        1364860800

        # Convert a time in a different timezone
        dt = flemming.convert_to_tz(
            datetime.datetime(2013, 4, 2, 4), pytz.timezone('US/Eastern'))
        print dt
        2013-04-02 00:00:00-04:00

        print flemming.unix_time(dt)
        1364875200

        # Print millisecond returns
        print flemming.unix_time(dt, return_ms=True)
        1364875200000

        # Do a unix_time conversion with respect to another timezone. When
        # it is converted back to a datetime, the time values are correct.
        # The original timezone, however, needs to be added back
        dt = datetime.datetime(2013, 2, 1, 5)

        # Print its EST time for later reference
        print flemming.convert_to_tz(dt, pytz.timezone('US/Eastern'))
        2013-02-01 00:00:00-05:00

        unix_tz_dt = flemming.unix_time(
            dt, within_tz=pytz.timezone('US/Eastern'))
        print unix_tz_dt
        1359676800

        # The datetime should match the original UTC time converted in
        # the timezone of the within_tz parameter. Tz information is
        # originally lost when converting to unix time, so replace the
        # tzinfo object here
        dt = datetime.datetime.fromtimestamp(unix_tz_dt).replace(
            tzinfo=pytz.timezone('US/Eastern'))
        print dt
        2013-02-01 00:00:00-05:00
    """
    epoch = datetime.datetime.utcfromtimestamp(0)
    offset = convert_to_tz(dt, within_tz).utcoffset().total_seconds() if within_tz else 0
    # Convert the timezone to UTC for arithmetic and make it naive
    dt = convert_to_tz(dt, pytz.utc).replace(tzinfo=None)
    unix_time = (dt - epoch).total_seconds() + offset
    return int(unix_time * 1000 if return_ms else unix_time)
